syntax = "proto2";

package interface.display;

import "common/proto/geometry.proto";

enum PrimitiveType {
  POINTS = 1;
  LINES = 2;
  TRIANGLES = 3;
  POLYGON = 4;
}

message Color {
  optional float red = 1;
  optional float green = 2;
  optional float blue = 3;
  optional float alpha = 4 [default = 1.0];
}

message RenderingState {
  optional Color color = 1;
  optional float line_width = 2;
}

message PrimitiveDrawCall {
  optional PrimitiveType type = 1;
  optional RenderingState rendering_state = 2;
  repeated interface.geometry.Point3D vertex = 3;
}

message VerticalSign {
  optional interface.geometry.Point3D center = 1;
  optional interface.geometry.Point2D size = 2;
  optional float angle = 3;
  optional string texture_name = 4;
}

// Generated from GL/gl.h
enum Function {
  //  GLCLEARINDEX = 0;
  //  GLCLEARCOLOR = 1;
  //  GLCLEAR = 2;
  //  GLINDEXMASK = 3;
  //  GLCOLORMASK = 4;
  //  GLALPHAFUNC = 5;
  //  GLBLENDFUNC = 6;
  //  GLLOGICOP = 7;
  //  GLCULLFACE = 8;
  //  GLFRONTFACE = 9;
  GLPOINTSIZE = 10;
  GLLINEWIDTH = 11;
  //  GLLINESTIPPLE = 12;
  //  GLPOLYGONMODE = 13;
  //  GLPOLYGONOFFSET = 14;
  //  GLPOLYGONSTIPPLE = 15;
  //  GLGETPOLYGONSTIPPLE = 16;
  //  GLEDGEFLAG = 17;
  //  GLEDGEFLAGV = 18;
  //  GLSCISSOR = 19;
  //  GLCLIPPLANE = 20;
  //  GLGETCLIPPLANE = 21;
  //  GLDRAWBUFFER = 22;
  //  GLREADBUFFER = 23;
  GLENABLE = 24;
  GLDISABLE = 25;
  //  GLISENABLED = 26;
  //  GLENABLECLIENTSTATE = 27;
  //  GLDISABLECLIENTSTATE = 28;
  //  GLGETBOOLEANV = 29;
  //  GLGETDOUBLEV = 30;
  //  GLGETFLOATV = 31;
  //  GLGETINTEGERV = 32;
  //  GLPUSHATTRIB = 33;
  //  GLPOPATTRIB = 34;
  //  GLPUSHCLIENTATTRIB = 35;
  //  GLPOPCLIENTATTRIB = 36;
  //  GLRENDERMODE = 37;
  //  GLGETERROR = 38;
  //  GLFINISH = 39;
  //  GLFLUSH = 40;
  //  GLHINT = 41;
  //  GLCLEARDEPTH = 42;
  //  GLDEPTHFUNC = 43;
  //  GLDEPTHMASK = 44;
  //  GLDEPTHRANGE = 45;
  //  GLCLEARACCUM = 46;
  //  GLACCUM = 47;
  //  GLMATRIXMODE = 48;
  //  GLORTHO = 49;
  //  GLFRUSTUM = 50;
  //  GLVIEWPORT = 51;
  GLPUSHMATRIX = 52;
  GLPOPMATRIX = 53;
  //  GLLOADIDENTITY = 54;
  //  GLLOADMATRIXD = 55;
  //  GLLOADMATRIXF = 56;
  //  GLMULTMATRIXD = 57;
  //  GLMULTMATRIXF = 58;
  //  GLROTATED = 59;
  GLROTATEF = 60;
  //  GLSCALED = 61;
  GLSCALEF = 62;
  //  GLTRANSLATED = 63;
  GLTRANSLATEF = 64;
  //  GLISLIST = 65;
  //  GLDELETELISTS = 66;
  //  GLGENLISTS = 67;
  //  GLNEWLIST = 68;
  //  GLENDLIST = 69;
  //  GLCALLLIST = 70;
  //  GLCALLLISTS = 71;
  //  GLLISTBASE = 72;
  GLBEGIN = 73;
  GLEND = 74;
  //  GLVERTEX2D = 75;
  //  GLVERTEX2F = 76;
  //  GLVERTEX2I = 77;
  //  GLVERTEX2S = 78;
  //  GLVERTEX3D = 79;
  GLVERTEX3F = 80;
  //  GLVERTEX3I = 81;
  //  GLVERTEX3S = 82;
  //  GLVERTEX4D = 83;
  //  GLVERTEX4F = 84;
  //  GLVERTEX4I = 85;
  //  GLVERTEX4S = 86;
  //  GLVERTEX2DV = 87;
  //  GLVERTEX2FV = 88;
  //  GLVERTEX2IV = 89;
  //  GLVERTEX2SV = 90;
  //  GLVERTEX3DV = 91;
  //  GLVERTEX3FV = 92;
  //  GLVERTEX3IV = 93;
  //  GLVERTEX3SV = 94;
  //  GLVERTEX4DV = 95;
  //  GLVERTEX4FV = 96;
  //  GLVERTEX4IV = 97;
  //  GLVERTEX4SV = 98;
  //  GLNORMAL3B = 99;
  //  GLNORMAL3D = 100;
  //  GLNORMAL3F = 101;
  //  GLNORMAL3I = 102;
  //  GLNORMAL3S = 103;
  //  GLNORMAL3BV = 104;
  //  GLNORMAL3DV = 105;
  //  GLNORMAL3FV = 106;
  //  GLNORMAL3IV = 107;
  //  GLNORMAL3SV = 108;
  //  GLINDEXD = 109;
  //  GLINDEXF = 110;
  //  GLINDEXI = 111;
  //  GLINDEXS = 112;
  //  GLINDEXUB = 113;
  //  GLINDEXDV = 114;
  //  GLINDEXFV = 115;
  //  GLINDEXIV = 116;
  //  GLINDEXSV = 117;
  //  GLINDEXUBV = 118;
  //  GLCOLOR3B = 119;
  //  GLCOLOR3D = 120;
  GLCOLOR3F = 121;
  //  GLCOLOR3I = 122;
  //  GLCOLOR3S = 123;
  //  GLCOLOR3UB = 124;
  //  GLCOLOR3UI = 125;
  //  GLCOLOR3US = 126;
  //  GLCOLOR4B = 127;
  //  GLCOLOR4D = 128;
  GLCOLOR4F = 129;
  //  GLCOLOR4I = 130;
  //  GLCOLOR4S = 131;
  //  GLCOLOR4UB = 132;
  //  GLCOLOR4UI = 133;
  //  GLCOLOR4US = 134;
  //  GLCOLOR3BV = 135;
  //  GLCOLOR3DV = 136;
  //  GLCOLOR3FV = 137;
  //  GLCOLOR3IV = 138;
  //  GLCOLOR3SV = 139;
  //  GLCOLOR3UBV = 140;
  //  GLCOLOR3UIV = 141;
  //  GLCOLOR3USV = 142;
  //  GLCOLOR4BV = 143;
  //  GLCOLOR4DV = 144;
  //  GLCOLOR4FV = 145;
  //  GLCOLOR4IV = 146;
  //  GLCOLOR4SV = 147;
  //  GLCOLOR4UBV = 148;
  //  GLCOLOR4UIV = 149;
  //  GLCOLOR4USV = 150;
  //  GLTEXCOORD1D = 151;
  //  GLTEXCOORD1F = 152;
  //  GLTEXCOORD1I = 153;
  //  GLTEXCOORD1S = 154;
  //  GLTEXCOORD2D = 155;
  //  GLTEXCOORD2F = 156;
  //  GLTEXCOORD2I = 157;
  //  GLTEXCOORD2S = 158;
  //  GLTEXCOORD3D = 159;
  //  GLTEXCOORD3F = 160;
  //  GLTEXCOORD3I = 161;
  //  GLTEXCOORD3S = 162;
  //  GLTEXCOORD4D = 163;
  //  GLTEXCOORD4F = 164;
  //  GLTEXCOORD4I = 165;
  //  GLTEXCOORD4S = 166;
  //  GLTEXCOORD1DV = 167;
  //  GLTEXCOORD1FV = 168;
  //  GLTEXCOORD1IV = 169;
  //  GLTEXCOORD1SV = 170;
  //  GLTEXCOORD2DV = 171;
  //  GLTEXCOORD2FV = 172;
  //  GLTEXCOORD2IV = 173;
  //  GLTEXCOORD2SV = 174;
  //  GLTEXCOORD3DV = 175;
  //  GLTEXCOORD3FV = 176;
  //  GLTEXCOORD3IV = 177;
  //  GLTEXCOORD3SV = 178;
  //  GLTEXCOORD4DV = 179;
  //  GLTEXCOORD4FV = 180;
  //  GLTEXCOORD4IV = 181;
  //  GLTEXCOORD4SV = 182;
  //  GLRASTERPOS2D = 183;
  //  GLRASTERPOS2F = 184;
  //  GLRASTERPOS2I = 185;
  //  GLRASTERPOS2S = 186;
  //  GLRASTERPOS3D = 187;
  //  GLRASTERPOS3F = 188;
  //  GLRASTERPOS3I = 189;
  //  GLRASTERPOS3S = 190;
  //  GLRASTERPOS4D = 191;
  //  GLRASTERPOS4F = 192;
  //  GLRASTERPOS4I = 193;
  //  GLRASTERPOS4S = 194;
  //  GLRASTERPOS2DV = 195;
  //  GLRASTERPOS2FV = 196;
  //  GLRASTERPOS2IV = 197;
  //  GLRASTERPOS2SV = 198;
  //  GLRASTERPOS3DV = 199;
  //  GLRASTERPOS3FV = 200;
  //  GLRASTERPOS3IV = 201;
  //  GLRASTERPOS3SV = 202;
  //  GLRASTERPOS4DV = 203;
  //  GLRASTERPOS4FV = 204;
  //  GLRASTERPOS4IV = 205;
  //  GLRASTERPOS4SV = 206;
  //  GLRECTD = 207;
  //  GLRECTF = 208;
  //  GLRECTI = 209;
  //  GLRECTS = 210;
  //  GLRECTDV = 211;
  //  GLRECTFV = 212;
  //  GLRECTIV = 213;
  //  GLRECTSV = 214;
  //  GLVERTEXPOINTER = 215;
  //  GLNORMALPOINTER = 216;
  //  GLCOLORPOINTER = 217;
  //  GLINDEXPOINTER = 218;
  //  GLTEXCOORDPOINTER = 219;
  //  GLEDGEFLAGPOINTER = 220;
  //  GLGETPOINTERV = 221;
  //  GLARRAYELEMENT = 222;
  //  GLDRAWARRAYS = 223;
  //  GLDRAWELEMENTS = 224;
  //  GLINTERLEAVEDARRAYS = 225;
  //  GLSHADEMODEL = 226;
  //  GLLIGHTF = 227;
  //  GLLIGHTI = 228;
  //  GLLIGHTFV = 229;
  //  GLLIGHTIV = 230;
  //  GLGETLIGHTFV = 231;
  //  GLGETLIGHTIV = 232;
  //  GLLIGHTMODELF = 233;
  //  GLLIGHTMODELI = 234;
  //  GLLIGHTMODELFV = 235;
  //  GLLIGHTMODELIV = 236;
  //  GLMATERIALF = 237;
  //  GLMATERIALI = 238;
  //  GLMATERIALFV = 239;
  //  GLMATERIALIV = 240;
  //  GLGETMATERIALFV = 241;
  //  GLGETMATERIALIV = 242;
  //  GLCOLORMATERIAL = 243;
  //  GLPIXELZOOM = 244;
  //  GLPIXELSTOREF = 245;
  //  GLPIXELSTOREI = 246;
  //  GLPIXELTRANSFERF = 247;
  //  GLPIXELTRANSFERI = 248;
  //  GLPIXELMAPFV = 249;
  //  GLPIXELMAPUIV = 250;
  //  GLPIXELMAPUSV = 251;
  //  GLGETPIXELMAPFV = 252;
  //  GLGETPIXELMAPUIV = 253;
  //  GLGETPIXELMAPUSV = 254;
  //  GLBITMAP = 255;
  //  GLREADPIXELS = 256;
  //  GLDRAWPIXELS = 257;
  //  GLCOPYPIXELS = 258;
  //  GLSTENCILFUNC = 259;
  //  GLSTENCILMASK = 260;
  //  GLSTENCILOP = 261;
  //  GLCLEARSTENCIL = 262;
  //  GLTEXGEND = 263;
  //  GLTEXGENF = 264;
  //  GLTEXGENI = 265;
  //  GLTEXGENDV = 266;
  //  GLTEXGENFV = 267;
  //  GLTEXGENIV = 268;
  //  GLGETTEXGENDV = 269;
  //  GLGETTEXGENFV = 270;
  //  GLGETTEXGENIV = 271;
  //  GLTEXENVF = 272;
  //  GLTEXENVI = 273;
  //  GLTEXENVFV = 274;
  //  GLTEXENVIV = 275;
  //  GLGETTEXENVFV = 276;
  //  GLGETTEXENVIV = 277;
  //  GLTEXPARAMETERF = 278;
  //  GLTEXPARAMETERI = 279;
  //  GLTEXPARAMETERFV = 280;
  //  GLTEXPARAMETERIV = 281;
  //  GLGETTEXPARAMETERFV = 282;
  //  GLGETTEXPARAMETERIV = 283;
  //  GLGETTEXLEVELPARAMETERFV = 284;
  //  GLGETTEXLEVELPARAMETERIV = 285;
  //  GLTEXIMAGE1D = 286;
  //  GLTEXIMAGE2D = 287;
  //  GLGETTEXIMAGE = 288;
  //  GLGENTEXTURES = 289;
  //  GLDELETETEXTURES = 290;
  //  GLBINDTEXTURE = 291;
  //  GLPRIORITIZETEXTURES = 292;
  //  GLARETEXTURESRESIDENT = 293;
  //  GLISTEXTURE = 294;
  //  GLTEXSUBIMAGE1D = 295;
  //  GLTEXSUBIMAGE2D = 296;
  //  GLCOPYTEXIMAGE1D = 297;
  //  GLCOPYTEXIMAGE2D = 298;
  //  GLCOPYTEXSUBIMAGE1D = 299;
  //  GLCOPYTEXSUBIMAGE2D = 300;
  //  GLMAP1D = 301;
  //  GLMAP1F = 302;
  //  GLMAP2D = 303;
  //  GLMAP2F = 304;
  //  GLGETMAPDV = 305;
  //  GLGETMAPFV = 306;
  //  GLGETMAPIV = 307;
  //  GLEVALCOORD1D = 308;
  //  GLEVALCOORD1F = 309;
  //  GLEVALCOORD1DV = 310;
  //  GLEVALCOORD1FV = 311;
  //  GLEVALCOORD2D = 312;
  //  GLEVALCOORD2F = 313;
  //  GLEVALCOORD2DV = 314;
  //  GLEVALCOORD2FV = 315;
  //  GLMAPGRID1D = 316;
  //  GLMAPGRID1F = 317;
  //  GLMAPGRID2D = 318;
  //  GLMAPGRID2F = 319;
  //  GLEVALPOINT1 = 320;
  //  GLEVALPOINT2 = 321;
  //  GLEVALMESH1 = 322;
  //  GLEVALMESH2 = 323;
  //  GLFOGF = 324;
  //  GLFOGI = 325;
  //  GLFOGFV = 326;
  //  GLFOGIV = 327;
  //  GLFEEDBACKBUFFER = 328;
  //  GLPASSTHROUGH = 329;
  //  GLSELECTBUFFER = 330;
  //  GLINITNAMES = 331;
  //  GLLOADNAME = 332;
  //  GLPUSHNAME = 333;
  //  GLPOPNAME = 334;
  //  GLDRAWRANGEELEMENTS = 335;
  //  GLTEXIMAGE3D = 336;
  //  GLTEXSUBIMAGE3D = 337;
  //  GLCOPYTEXSUBIMAGE3D = 338;
  //  GLCOLORTABLE = 339;
  //  GLCOLORSUBTABLE = 340;
  //  GLCOLORTABLEPARAMETERIV = 341;
  //  GLCOLORTABLEPARAMETERFV = 342;
  //  GLCOPYCOLORSUBTABLE = 343;
  //  GLCOPYCOLORTABLE = 344;
  //  GLGETCOLORTABLE = 345;
  //  GLGETCOLORTABLEPARAMETERFV = 346;
  //  GLGETCOLORTABLEPARAMETERIV = 347;
  //  GLBLENDEQUATION = 348;
  //  GLBLENDCOLOR = 349;
  //  GLHISTOGRAM = 350;
  //  GLRESETHISTOGRAM = 351;
  //  GLGETHISTOGRAM = 352;
  //  GLGETHISTOGRAMPARAMETERFV = 353;
  //  GLGETHISTOGRAMPARAMETERIV = 354;
  //  GLMINMAX = 355;
  //  GLRESETMINMAX = 356;
  //  GLGETMINMAX = 357;
  //  GLGETMINMAXPARAMETERFV = 358;
  //  GLGETMINMAXPARAMETERIV = 359;
  //  GLCONVOLUTIONFILTER1D = 360;
  //  GLCONVOLUTIONFILTER2D = 361;
  //  GLCONVOLUTIONPARAMETERF = 362;
  //  GLCONVOLUTIONPARAMETERFV = 363;
  //  GLCONVOLUTIONPARAMETERI = 364;
  //  GLCONVOLUTIONPARAMETERIV = 365;
  //  GLCOPYCONVOLUTIONFILTER1D = 366;
  //  GLCOPYCONVOLUTIONFILTER2D = 367;
  //  GLGETCONVOLUTIONFILTER = 368;
  //  GLGETCONVOLUTIONPARAMETERFV = 369;
  //  GLGETCONVOLUTIONPARAMETERIV = 370;
  //  GLSEPARABLEFILTER2D = 371;
  //  GLGETSEPARABLEFILTER = 372;
  //  GLACTIVETEXTURE = 373;
  //  GLCLIENTACTIVETEXTURE = 374;
  //  GLCOMPRESSEDTEXIMAGE1D = 375;
  //  GLCOMPRESSEDTEXIMAGE2D = 376;
  //  GLCOMPRESSEDTEXIMAGE3D = 377;
  //  GLCOMPRESSEDTEXSUBIMAGE1D = 378;
  //  GLCOMPRESSEDTEXSUBIMAGE2D = 379;
  //  GLCOMPRESSEDTEXSUBIMAGE3D = 380;
  //  GLGETCOMPRESSEDTEXIMAGE = 381;
  //  GLMULTITEXCOORD1D = 382;
  //  GLMULTITEXCOORD1DV = 383;
  //  GLMULTITEXCOORD1F = 384;
  //  GLMULTITEXCOORD1FV = 385;
  //  GLMULTITEXCOORD1I = 386;
  //  GLMULTITEXCOORD1IV = 387;
  //  GLMULTITEXCOORD1S = 388;
  //  GLMULTITEXCOORD1SV = 389;
  //  GLMULTITEXCOORD2D = 390;
  //  GLMULTITEXCOORD2DV = 391;
  //  GLMULTITEXCOORD2F = 392;
  //  GLMULTITEXCOORD2FV = 393;
  //  GLMULTITEXCOORD2I = 394;
  //  GLMULTITEXCOORD2IV = 395;
  //  GLMULTITEXCOORD2S = 396;
  //  GLMULTITEXCOORD2SV = 397;
  //  GLMULTITEXCOORD3D = 398;
  //  GLMULTITEXCOORD3DV = 399;
  //  GLMULTITEXCOORD3F = 400;
  //  GLMULTITEXCOORD3FV = 401;
  //  GLMULTITEXCOORD3I = 402;
  //  GLMULTITEXCOORD3IV = 403;
  //  GLMULTITEXCOORD3S = 404;
  //  GLMULTITEXCOORD3SV = 405;
  //  GLMULTITEXCOORD4D = 406;
  //  GLMULTITEXCOORD4DV = 407;
  //  GLMULTITEXCOORD4F = 408;
  //  GLMULTITEXCOORD4FV = 409;
  //  GLMULTITEXCOORD4I = 410;
  //  GLMULTITEXCOORD4IV = 411;
  //  GLMULTITEXCOORD4S = 412;
  //  GLMULTITEXCOORD4SV = 413;
  //  GLLOADTRANSPOSEMATRIXD = 414;
  //  GLLOADTRANSPOSEMATRIXF = 415;
  //  GLMULTTRANSPOSEMATRIXD = 416;
  //  GLMULTTRANSPOSEMATRIXF = 417;
  //  GLSAMPLECOVERAGE = 418;
  //  GLACTIVETEXTUREARB = 419;
  //  GLCLIENTACTIVETEXTUREARB = 420;
  //  GLMULTITEXCOORD1DARB = 421;
  //  GLMULTITEXCOORD1DVARB = 422;
  //  GLMULTITEXCOORD1FARB = 423;
  //  GLMULTITEXCOORD1FVARB = 424;
  //  GLMULTITEXCOORD1IARB = 425;
  //  GLMULTITEXCOORD1IVARB = 426;
  //  GLMULTITEXCOORD1SARB = 427;
  //  GLMULTITEXCOORD1SVARB = 428;
  //  GLMULTITEXCOORD2DARB = 429;
  //  GLMULTITEXCOORD2DVARB = 430;
  //  GLMULTITEXCOORD2FARB = 431;
  //  GLMULTITEXCOORD2FVARB = 432;
  //  GLMULTITEXCOORD2IARB = 433;
  //  GLMULTITEXCOORD2IVARB = 434;
  //  GLMULTITEXCOORD2SARB = 435;
  //  GLMULTITEXCOORD2SVARB = 436;
  //  GLMULTITEXCOORD3DARB = 437;
  //  GLMULTITEXCOORD3DVARB = 438;
  //  GLMULTITEXCOORD3FARB = 439;
  //  GLMULTITEXCOORD3FVARB = 440;
  //  GLMULTITEXCOORD3IARB = 441;
  //  GLMULTITEXCOORD3IVARB = 442;
  //  GLMULTITEXCOORD3SARB = 443;
  //  GLMULTITEXCOORD3SVARB = 444;
  //  GLMULTITEXCOORD4DARB = 445;
  //  GLMULTITEXCOORD4DVARB = 446;
  //  GLMULTITEXCOORD4FARB = 447;
  //  GLMULTITEXCOORD4FVARB = 448;
  //  GLMULTITEXCOORD4IARB = 449;
  //  GLMULTITEXCOORD4IVARB = 450;
  //  GLMULTITEXCOORD4SARB = 451;
  //  GLMULTITEXCOORD4SVARB = 452;
  //  GLBLENDEQUATIONSEPARATEATI = 453;
  //  GLEGLIMAGETARGETTEXTURE2DOES = 454;
  //  GLEGLIMAGETARGETRENDERBUFFERSTORAGEOES = 455;
}

message Parameter {
  oneof parameter {
    float float_parameter = 1; // for GLfloat
    sfixed32 int32_parameter = 2; // for GLenum, GLint
  }
}

message Canvas {
  optional string name = 2;
  repeated PrimitiveDrawCall primitive_draw_call = 4;
  repeated VerticalSign vertical_sign = 5;
}

message CanvasList {
  repeated Canvas canvas = 1;
}

message Variable {
  optional string name = 1;
  optional string value = 2;
  optional Color foreground_color = 3;
  optional Color background_color = 4;
}

message VariableView {
  optional string name = 2;
  repeated Variable variable = 3;
}

message VariableViewList {
  repeated VariableView variable_view = 1;
}
